nbfetch/__init__.py

```diff
@@ -5,32 +5,80 @@
 from tornado.web import StaticFileHandler
 import os
 
+import jupyter_server
+from jupyter_server.extension.application import ExtensionApp
+from notebook import DEFAULT_STATIC_FILES_PATH, DEFAULT_TEMPLATE_PATH_LIST
+import jinja2
+import gettext
 
-def _jupyter_server_extension_paths():
-    return [{
-        'module': 'nbfetch',
-    }]
-
-
-def load_jupyter_server_extension(nbapp):
-    web_app = nbapp.web_app
-    base_url = url_path_join(web_app.settings['base_url'], 'git-pull')
-    hs_url = url_path_join(web_app.settings['base_url'], 'hs-pull')
-    handlers = [
-        (url_path_join(base_url, 'api'), SyncHandler),
-        (url_path_join(hs_url, 'api'), HSyncHandler),
-        (base_url, UIHandler),
-        (hs_url, HSHandler),
-        (
-            url_path_join(base_url, 'static', '(.*)'),
-            StaticFileHandler,
-            {'path': os.path.join(os.path.dirname(__file__), 'static')}
-        ),
-        (
-            url_path_join(hs_url, 'static', '(.*)'),
-            StaticFileHandler,
-            {'path': os.path.join(os.path.dirname(__file__), 'static')}
-        ),
-        (url_path_join(web_app.settings['base_url'], 'hslogin'), HSLoginHandler),
+
+HERE = os.path.dirname(__file__)
+
+
+class NbFetchApp(ExtensionApp):
+    # Name of the extension.
+    name = "nbfetch"
+    default_url = "/"
+    load_other_extensions = True
+    file_url_prefix = "/"
+
+    # Local path to static files directory.
+    static_paths = [
+        os.path.join(HERE, "static"),
+        DEFAULT_STATIC_FILES_PATH,
     ]
-    web_app.add_handlers('.*', handlers)
+
+    def initialize_templates(self):
```

Comment:
Not sure where you pulled this from (assuming you copied it from a template and didn't do it yourself) but it looks like jupyter-server provides a mixin for jinja which might also fit your needs.  https://github.com/jupyter-server/jupyter_server/blob/main/jupyter_server/extension/application.py#L77

Reply:
Nice catch, I think... I tried that by specifying `template_paths` as an object property initially and ran into issues and I think in trying to debug something else, I ended up overridding `initialize_templates`. Ill try removing this implementation and see if I run into issues.

Reply:
So, when I remove `initialize_templates`, I get a `i18n` related exception thrown by `jinja2` (see below).

```shell
[E 2022-07-25 13:09:27.375 ServerApp] Uncaught exception GET /hs-pull?id=91dbf553b3f241d590d020997ec01ff5 (127.0.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='GET', uri='/hs-pull?id=91dbf553b3f241d590d020997ec01ff5', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "~/github/nbfetch/venv/lib/python3.9/site-packages/tornado/web.py", line 1713, in _execute
        result = await result
      File "~/github/nbfetch/venv/lib/python3.9/site-packages/tornado/gen.py", line 216, in wrapper
        result = ctx_run(func, *args, **kwargs)
      File "~/github/nbfetch/nbfetch/handlers.py", line 407, in get
        self.render_template("hstatus.html", id=id, path=path, version=__version__)
      File "~/github/nbfetch/venv/lib/python3.9/site-packages/jupyter_server/base/handlers.py", line 544, in render_template
        return template.render(**ns)
      File "~/github/nbfetch/venv/lib/pthon3.9/site-packages/jinja2/environment.py", line 1301, in render
        self.environment.handle_exception()
      File "~/github/nbfetch/venv/lib/python3.9/site-packages/jinja2/environment.py", line 936, in handle_exception
        raise rewrite_traceback_stack(source=source)
      File "~/github/nbfetch/nbfetch/templates/hstatus.html", line 1, in top-level template code
        {% extends "page.html" %}
      File "~/github/nbfetch/venv/lib/python3.9/site-packages/notebook/templates/page.html", line 120, in top-level template code
        {% trans %}Jupyter Notebook requires JavaScript.{% endtrans %}<br>
      File "~/github/nbfetch/venv/lib/python3.9/site-packages/jinja2/utils.py", line 83, in from_obj
        if hasattr(obj, "jinja_pass_arg"):
    jinja2.exceptions.UndefinedError: 'gettext' is undefinedy
```

Reply:
>The mixin classes in this example must come before the base classes, ExtensionApp and ExtensionHandler.

If that isn't your issue, I'm out of ideas. https://jupyter-server.readthedocs.io/en/latest/developers/extensions.html#jinja-templating-from-frontend-extensions

Reply:
Sorry, I should have mentioned that I received the above exception while including the `jupyter_server.extension.application.ExtensionAppJinjaMixin` mixin in `NbFetchApp`.

Reply:
Removing the mixing and results in a different error (`KeyError` related to finding the template jinja2 environment).

Reply:
> Sorry, I should have mentioned that I received the above exception while including the jupyter_server.extension.application.ExtensionAppJinjaMixin mixin in NbFetchApp.

Right, it was just a note that the order matters when including the mixin.

Correct:
`class NbFetchApp(ExtensionAppJinjaMixin, ExtensionApp):`

Incorrect:
`class NbFetchApp(ExtensionApp, ExtensionAppJinjaMixin):

Reply:
Right, I used the correct mro when I tried the above.